<a href="https://colab.research.google.com/github/GeoLabUniLaSalle/Python/blob/main/Buffet_%C3%A0_volont%C3%A9.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **Buffet à volonté ★☆☆**

*Exercice disponible sur la plateforme TOSA : Battle Dev RegionJob - Mars 2018 : https://www.isograd.com/FR/solutionconcours.php*

**Énoncé**

L'informatique c'est bien, mais la cuisine c'est encore mieux. Lassé de quitter vos cantines préférées en ayant encore faim, vous avez décidé de monter un restaurant avec un concept de buffet à volonté. Comme vous voyez grand, vous avez décidé de faire des "réductions de groupe" :
- 10% sur la note pour les groupes d'au moins 4 personnes
- 20% sur la note pour les groupes d'au moins 6 personnes
- 30% sur la note pour les groupes d'au moins 10 personnes


Pour votre business plan, vous souhaitez estimer le montant des ventes en fonction du prix par personne et de la configuration des tables servies.



**Format des données**

*Entrée*

Ligne 1 : un entier compris entre 20 et 40 correspondant au prix du buffet par personne.
Ligne 2 : un entier N compris entre 5 et 50 correspondant au nombre de tables servies.
Lignes 3 à N+2 : un entier compris entre 2 et 15 représentant le nombre de personnes présentes à une table.

*Sortie*

Un entier représentant le montant des ventes du restaurant arrondi à l'entier supérieur.

## **Données**

*Accès aux données*

In [ ]:
import csv
filedir = ('https://cours.unilasalle.fr/courses/SG4007/document/Exercices_Python/Buffet_a_volonte/')

import ssl
ssl._create_default_https_context = ssl._create_unverified_context

import urllib.request

for x in range(1,4):
  f = urllib.request.urlopen(str(filedir+"input"+str(x)+".txt"))

  data=[]
  for i in f:
      line=i.decode('utf-8').strip()
      data.append(line)
  print('Entrée',x,':',data)

  f = urllib.request.urlopen(str(filedir+"output"+str(x)+".txt"))
  for i in f:
      line=i.decode('utf-8').strip()
  print('Sortie',x,':',line,'\n')

Entrée 1 : ['37', '7', '3', '15', '14', '5', '5', '14', '3']
Sortie 1 : 1669 

Entrée 2 : ['21', '9', '13', '14', '12', '14', '8', '5', '15', '2', '10']
Sortie 2 : 1418 

Entrée 3 : ['35', '22', '3', '8', '12', '3', '7', '5', '6', '14', '4', '12', '10', '5', '2', '14', '11', '5', '13', '13', '6', '7', '4', '10']
Sortie 3 : 4627 



## **Traitement**

Reprenons de jeu de données 1 :

In [ ]:
lines = ['37', '7', '3', '15', '14', '5', '5', '14', '3']
print(lines)

['37', '7', '3', '15', '14', '5', '5', '14', '3']



Ici, nous avons le prix du menu (37 €), le nombre de tables (7 tables) et le nombre de convives pour chacune des 7 tables du restaurant.

La première table accueille 3 convives.
La seconde table accueille 15 convives.
etc.

Pour la première table :
3 convives prennent chacun un menu à 37 € et n'obtiennent pas de remise parce qu'ils sont moins de 4 personnes.

3 x 37 = 111 €

Pour la seconde table :
15 convives prennent chacun un menu à 37 € et obtiennent une remise de 30% parce qu'ils sont au moins 10 personnes.

15 x 37 x 0.7 = 388.5 €

Nous réalisons cette opération pour chacune des tables et en faisant la somme des additions, nous obtenons la valeur de sortie attendue : 1669 € (arrondi à l'entier supérieur)



In [ ]:
prix = int(lines[0])
print(prix)

37


Parcourons maintenant les nombres de personnes par table (On commence à parcourir à partir de l'élément d'indice 2 pour ne pas traiter les 2 premiers éléments qui contiennent le prix du menu et le nombre de tables) :

In [ ]:
for x in lines[2:]:
  print(x)

3
15
14
5
5
14
3


Obtenons une valeur entière sur laquelle nous pourrons faire un calcul :

In [ ]:
for x in lines[2:]:
  nb = int(x)
  print(nb)

3
15
14
5
5
14
3


Calculons maintenant la valeur de la réduction en fonction du nombre de personnes à table ;


In [ ]:
for x in lines[2:]:
  nb = int(x)
  if(nb>9):
    reduc = 0.7
  elif(nb>5):
    reduc = 0.8
  elif(nb>3):
    reduc = 0.9
  else:
    reduc = 1
  print(nb,reduc)

3 1
15 0.7
14 0.7
5 0.9
5 0.9
14 0.7
3 1


Calculons l'addition pour chaque table :

In [ ]:
for x in lines[2:]:
  nb = int(x)
  if(nb>9):
    reduc = 0.7
  elif(nb>5):
    reduc = 0.8
  elif(nb>3):
    reduc = 0.9
  else:
    reduc = 1
  print(nb * prix * reduc)

111
388.5
362.59999999999997
166.5
166.5
362.59999999999997
111


Faisons la somme de ces additions :

In [ ]:
somme = 0
for x in lines[2:]:
  nb = int(x)
  if(nb>9):
    reduc = 0.7
  elif(nb>5):
    reduc = 0.8
  elif(nb>3):
    reduc = 0.9
  else:
    reduc = 1
  somme += nb * prix * reduc
print(somme)

1668.6999999999998


Enfin, réalisons un arrondi à l'entier supérieur :

In [ ]:
import math as m
somme = 0
for x in lines[2:]:
  nb = int(x)
  if(nb>9):
    reduc = 0.7
  elif(nb>5):
    reduc = 0.8
  elif(nb>3):
    reduc = 0.9
  else:
    reduc = 1
  somme += nb * prix * reduc
print(m.ceil(somme))

1669


1669 est bien le résultat attendu !

Et voici un corrigé plus concis.

In [ ]:
from math import ceil

prix = int(lines[0])
nb_personnes = [int(a) for a in lines[2:]]

def note(prix, k):
    x = prix*k
    if k >= 10:
        x *= 0.7
    elif k >= 6:
        x *= 0.8
    elif k >= 4:
        x *= 0.9
    return x

print(ceil(sum(note(prix, k) for k in nb_personnes)))

1669
